core/src/Kokkos_WorkGraphPolicy.hpp

```diff
@@ -0,0 +1,210 @@
+#ifndef KOKKOS_WORKGRAPHPOLICY_HPP
+#define KOKKOS_WORKGRAPHPOLICY_HPP
+
+namespace Kokkos {
+namespace Experimental {
+
+template< class ... Properties >
+class WorkGraphPolicy
+{
+public:
+  using self_type = WorkGraphPolicy<Properties ... >;
+  using traits = Kokkos::Impl::PolicyTraits<Properties ... >;
+  using index_type = typename traits::index_type;
+  using execution_space = typename traits::execution_space;
+  using work_tag = typename traits::work_tag;
+  using memory_space = typename execution_space::memory_space;
+  using graph_type = Kokkos::Experimental::Crs<index_type, execution_space, void, index_type>;
+  using member_type = index_type;
+
+  graph_type graph;
+
+  WorkGraphPolicy(graph_type arg_graph):graph(arg_graph) {}
+};
+
+namespace Impl {
+
+template< class functor_type , class execution_space, class ... policy_args >
+class WorkGraphExec
+{
+ private:
+  struct TrackingReenabler {
+    TrackingReenabler() {
+      Kokkos::Impl::shared_allocation_tracking_enable();
+    }
+  } m_tracking_reenabler;
+ public:
+
+  using self_type = WorkGraphExec< functor_type, execution_space, policy_args ... >;
+  using policy_type = Kokkos::Experimental::WorkGraphPolicy< policy_args ... >;
+  using member_type = typename policy_type::member_type;
+  using memory_space = typename execution_space::memory_space;
+
+ protected:
+
+  const functor_type m_functor;
+  const policy_type  m_policy;
+
+ private:
+
+  using ints_type = Kokkos::View<std::int32_t*, memory_space>;
+  using atomic_ints_type = Kokkos::View<std::int32_t*, memory_space, MemoryTraits<Atomic>>;
+  using range_type = Kokkos::pair<std::int32_t, std::int32_t>;
+  using ranges_type = Kokkos::View<range_type*, memory_space>;
+  const std::int32_t m_total_work;
+  atomic_ints_type m_counts;
```

Comment:
What about atomic load and store? Cast pointer to volatile and then do load/store?